gax/src/test/java/com/google/api/gax/batching/BatcherImplTest.java

```diff
@@ -711,6 +712,134 @@ public void run() {
+    try {
+      assertThat(batcher2.getFlowController()).isSameInstanceAs(flowController);
+    } finally {
+      batcher2.close();
+    }
+  }
+
+  @Test
+  public void testThrottlingBlocking() throws Exception {
+    BatchingSettings settings =
+        BatchingSettings.newBuilder()
+            .setElementCountThreshold(1L)
+            .setRequestByteThreshold(1L)
+            .build();
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .setMaxOutstandingElementCount(1L)
+                .build());
+    final Batcher<Integer, Integer> batcher =
+        new BatcherImpl<>(
+            SQUARER_BATCHING_DESC_V2,
+            callLabeledIntSquarer,
+            labeledIntList,
+            settings,
+            EXECUTOR,
+            flowController);
+    ExecutorService executor = Executors.newSingleThreadExecutor();
+    try {
+      flowController.reserve(1, 1);
+      Future future =
+          executor.submit(
+              new Runnable() {
+                @Override
+                public void run() {
+                  batcher.add(1);
+                }
+              });
+      try {
+        future.get(100, TimeUnit.MILLISECONDS);
+        assertWithMessage("adding elements to batcher should be blocked by FlowControlled").fail();
+      } catch (TimeoutException e) {
+        // expected
+      }
+      flowController.release(1, 1);
+      try {
+        future.get(100, TimeUnit.MILLISECONDS);
+      } catch (TimeoutException e) {
+        assertWithMessage("adding elements to batcher should not be blocked").fail();
+      }
+    } finally {
+      batcher.close();
+      executor.shutdownNow();
+    }
+  }
+
+  @Test
+  public void testThrottlingNonBlocking() throws Exception {
+    BatchingSettings settings =
+        BatchingSettings.newBuilder()
+            .setElementCountThreshold(1L)
+            .setRequestByteThreshold(1L)
+            .build();
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setLimitExceededBehavior(LimitExceededBehavior.ThrowException)
+                .setMaxOutstandingElementCount(1L)
+                .build());
+    final Batcher<Integer, Integer> batcher =
+        new BatcherImpl<>(
+            SQUARER_BATCHING_DESC_V2,
+            callLabeledIntSquarer,
+            labeledIntList,
+            settings,
+            EXECUTOR,
+            flowController);
+    try {
+      flowController.reserve(1, 1);
+      try {
+        batcher.add(1);
+        assertWithMessage("Should throw exception because it exceeded FlowController limit").fail();
+      } catch (Exception e) {
+        assertThat(e.getMessage()).contains("The maximum number of batch elements");
+      }
+      flowController.release(1, 1);
+      batcher.add(1);
+    } finally {
```

Comment:
I think you can use try with resources for this since Batcher implements AutoCloseable